File: Design-Patterns/mvp/app/src/main/java/com/example/mvc/data/products/datasource/local/ProductsLocalDatasource.java
package com.example.mvc.data.products.datasource.local;

import android.content.Context;
import com.example.mvc.data.db.AppDatabase;
import com.example.mvc.data.products.model.Product;
import java.util.List;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Flowable;

public class ProductsLocalDatasource {
    private ProductDAO productDAO;

    public ProductsLocalDatasource(Context context){
        productDAO = AppDatabase.getInstance(context).productDAO();
    }

    public Completable insertProduct(Product product){
        return productDAO.insertProduct(product);
    }

    public Completable deleteProduct(Product product){
        return productDAO.deleteProduct(product);
    }

    public Flowable<List<Product>> getFavoriteProducts(){
        return productDAO.getFavoriteProducts();
    }
}

File: Design-Patterns/MVVM/app/src/main/java/com/example/mvvm/data/products/datasource/local/ProductsLocalDatasource.kt
package com.example.mvvm.data.products.datasource.local

import android.content.Context
import com.example.mvvm.data.db.AppDatabase
import com.example.mvvm.data.products.model.Product
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Flowable
import kotlinx.coroutines.flow.Flow

class ProductsLocalDatasource(context: Context) {
    private val productDao = AppDatabase.getInstance(context).productDao()

    fun getFavoriteProducts(): Flow<List<Product>> {
        return productDao.getAllProducts()
    }

    suspend fun insertProduct(product: Product) {
        productDao.insertProduct(product)
    }

    suspend fun deleteProduct(product: Product) {
        productDao.deleteProduct(product)
    }}